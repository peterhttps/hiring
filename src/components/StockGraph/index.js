import React, { useEffect, useState } from "react";
import {
  LineChart,
  Line,
  XAxis,
  YAxis,
  Tooltip,
  Legend,
  ResponsiveContainer,
} from "recharts";
import { Wrapper, ContainerDate } from "./styles";
import { DateButton } from "../Global/styles";

import DatePicker from "react-datepicker";

import "react-datepicker/dist/react-datepicker.css";

import { registerLocale } from "react-datepicker";
import us from "date-fns/locale/en-US";
import { ActionButton } from "../Global/styles";
registerLocale("en-US", us);

function StockGraph({ data }) {
  const [graphData, setGraphData] = useState([]);
  const [loading, setLoading] = useState(true);
  const [dateFromGlobal, setDateFromGlobal] = useState(new Date());
  const [dateToGlobal, setDateToGlobal] = useState(new Date());

  const ref = React.createRef();

  useEffect(() => {
    const dataStock = [];

    const days = Object.keys(data);
    var i = 0;
    for (i; i < 30; i++) {
      const values = data[days[i]];
      const dataCell = {
        date: days[i],
        price: values["4. close"],
      };
      dataStock.push(dataCell);
    }

    setGraphData(dataStock.reverse());
    setLoading(false);
  }, [data]);

  function changeData() {
    const days = Object.keys(data);

    const dateFrom = dateFromGlobal.toISOString().slice(0, 10);
    const dateTo = dateToGlobal.toISOString().slice(0, 10);
    var from = 0;
    var to = 0;

    for (var i = 0; i < days.length - 1; i++) {
      if (days[i] === dateFrom) {
        from = i;
      }
      if (days[i] === dateTo) {
        to = i;
      }
    }

    var dateStock = [];
    for (var toD = to; toD <= from; toD++) {
      const values = data[days[toD]];
      const dataCell = {
        date: days[toD],
        price: values["4. close"],
      };
      dateStock.push(dataCell);
    }
    setGraphData(dateStock.reverse());
  }

  const CustomDatePicker = React.forwardRef(({ value, onClick }, ref) => (
    <DateButton onClick={onClick} ref={ref}>
      {value}
    </DateButton>
  ));

  return (
    <Wrapper>
      {!loading && (
        <ResponsiveContainer width="99%" aspect={3}>
          <LineChart
            width={1200}
            height={300}
            data={graphData}
            margin={{
              top: 5,
              right: 80,
              left: 10,
              bottom: 5,
            }}
          >
            <XAxis dataKey="date" />
            <YAxis />
            <Tooltip contentStyle={{ color: "#000000" }} />
            <Legend />
            <Line
              type="monotone"
              dataKey="price"
              stroke="#8884d8"
              activeDot={{ r: 8 }}
            />
          </LineChart>
        </ResponsiveContainer>
      )}

      <ContainerDate>
        <p>From: </p>
        <DatePicker
          selected={dateFromGlobal}
          onChange={(date) => setDateFromGlobal(date)}
          locale="en-US"
          dateFormat="yyyy-MM-dd"
          customInput={<CustomDatePicker />}
          ref={ref}
        />
        <p>To: </p>
        <DatePicker
          selected={dateToGlobal}
          onChange={(date) => setDateToGlobal(date)}
          locale="en-US"
          dateFormat="yyyy-MM-dd"
          customInput={<CustomDatePicker />}
          ref={ref}
        />
        <ActionButton onClick={() => changeData()}>Filter</ActionButton>
      </ContainerDate>
    </Wrapper>
  );
}

export default StockGraph;
