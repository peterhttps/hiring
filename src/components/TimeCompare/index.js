import React, { useState } from "react";
import DatePicker from "react-datepicker";

import "react-datepicker/dist/react-datepicker.css";

import { Wrapper, Container, DatePickerContainer, ResultContainer, DateButton } from "./styles";

import { registerLocale } from  "react-datepicker";
import us from 'date-fns/locale/en-US';
import { ActionButton } from "../Global/styles";
registerLocale('en-US', us)

function TimeCompare({data}) {
    const [dateFromGlobal, setDateFromGlobal] = useState(new Date());
    const [dateToGlobal, setDateToGlobal] = useState(new Date());
    const [finalValue, setFinalValue] = useState("");

    const ref = React.createRef();

    const CustomDatePicker = React.forwardRef(({ value, onClick }, ref) => (
      <DateButton onClick={onClick} ref={ref}>{value}</DateButton>
  ));

    function timeCompare() {
      const days = Object.keys(data);

      const dateFrom = dateFromGlobal.toISOString().slice(0, 10);
      const dateTo = dateToGlobal.toISOString().slice(0, 10);
      var from = 0;
      var to = 0;

      for (var i = 0; i < days.length - 1; i++) {
        // console.log(days[i]);
        if (days[i] === dateFrom) {
            console.log("eh igua");
            from = i;
        }
        if (days[i] === dateTo) {
            console.log("eh igua tb");
            to = i;
        }
      } 

      const dateFromClose = parseFloat(data[days[from]]["4. close"]).toFixed(2);
      const dateToClose = parseFloat(data[days[to]]["4. close"]).toFixed(2);

      const result = dateToClose - dateFromClose;
      setFinalValue(result.toFixed(2));
    }



    return (
        <Wrapper>
          <h1>Time Compare</h1>
          <hr />
          <Container>
            <DatePickerContainer>
            <DatePicker
                selected={dateFromGlobal}
                onChange={(date) => setDateFromGlobal(date)}
                dateFormat="yyyy-MM-dd"
                locale="en-US"
                customInput={<CustomDatePicker />}
                ref={ref}
            />
            <DatePicker
                selected={dateToGlobal}
                onChange={(date) => setDateToGlobal(date)}
                dateFormat="yyyy-MM-dd"
                locale="en-US"
                customInput={<CustomDatePicker />}
                ref={ref}
            />
            <ActionButton onClick={() => timeCompare()}>Calculate</ActionButton>
            </DatePickerContainer>
            <ResultContainer>
              <h1>$ {finalValue}</h1>
            </ResultContainer>
            
            </Container>
        </Wrapper>
    );
}

export default TimeCompare;
