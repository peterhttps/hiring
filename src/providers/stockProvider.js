import React, { useEffect, useState } from 'react';


export const StockContext = React.createContext({});


export const StockProvider = (props) => {
    const [stocks, setStocks] = useState(JSON.parse(localStorage.getItem('stocks')) || []);

    useEffect(() => {
        localStorage.setItem('stocks', JSON.stringify(stocks));
    }, [stocks]);

    return (
        <StockContext.Provider value={{stocks, setStocks}}>
            {props.children}
        </StockContext.Provider>
    );
} 