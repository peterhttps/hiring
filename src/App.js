import { ThemeProvider } from "styled-components";
import { GlobalStyles } from "./themes/GlobalStyle";
import Routes from "./routes";
import { ThemeContext } from "./providers/themeProvider";
import { useContext } from "react";
import './styles.css';

function App(props) {
  const {theme} = useContext(ThemeContext);

  return (
    <ThemeProvider theme={theme}>
      <GlobalStyles />
      <Routes />
    </ThemeProvider>
  );
}

export default App;
