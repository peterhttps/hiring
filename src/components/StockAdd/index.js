import React, { useContext, useEffect, useState } from 'react';
import { StockContext } from '../../providers/stockProvider';

import { Wrapper } from './styles';
import { ActionButton } from '../Global/styles';

function StockAdd({actualStock}) {
  const {stocks, setStocks} = useContext(StockContext);
  const [isPresent, setIsPresent] = useState(false);
  
  useEffect(() => {
    if (stocks !== null) {
    stocks.forEach(element => {
      if (element.symbol === actualStock) {
        setIsPresent(true);
      }  
    });
  }
  }, [stocks, actualStock])

  function addStock() {
    setStocks((stocks) => [
      ...stocks,
      {
        symbol: actualStock
      },
    ]);
    setIsPresent(true);
  }

  function removeStock() {
    setStocks(stocks.filter(item => item.symbol !== actualStock));
    setIsPresent(false);
  }

  return (
      <Wrapper>
         {isPresent 
         ? <ActionButton onClick={() => removeStock()}>Remove from Portfolio</ActionButton>
         : <ActionButton onClick={() => addStock()}>Add to Portfolio</ActionButton>
         }
      </Wrapper>
  );
}

export default StockAdd;