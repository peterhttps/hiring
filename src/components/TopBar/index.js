import React, { useContext, useEffect, useState } from 'react';
import { ThemeContext } from '../../providers/themeProvider';

import { ContainerTopBar } from './styles';
import { lightTheme, darkTheme } from '../../themes/Themes';

import {FiSun, FiMoon} from 'react-icons/fi';
import { useHistory } from 'react-router-dom';

function TopBar() {
  const history = useHistory();

  const {theme, setTheme} = useContext(ThemeContext);
  const [isLight, setIsLight] = useState(true);

  useEffect(() => {
    if (theme?.name === "light") {
      setIsLight(true);
    } else {
      setIsLight(false);
    }
  }, [isLight, theme?.name]);

  function modeClick() {
    if (theme.name === "light") {
      setTheme(darkTheme);
    } else {
      setTheme(lightTheme);
    }
  }

  return (
    <ContainerTopBar>
        <button onClick={() => modeClick()}>
          {!isLight
          ? <FiSun color="#FFFFFF" size={24}/>
          : <FiMoon color="#FFFFFF" size={24}/>}
          </button>
        <div onClick={() => history.push("/portfolio")}>
          <h2>My Portf√≥lio</h2>
        </div>
    </ContainerTopBar>
  );
}

export default TopBar;