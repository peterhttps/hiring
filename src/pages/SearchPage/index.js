import React, { useState } from 'react';

import { Wrapper, Container, WrapperSearch, ContainerSearch } from './styles';
import TopBar from '../../components/TopBar';
import { useHistory } from 'react-router-dom';

function SearchPage() {
    const history = useHistory();
    
    const [stock, setStock] = useState("");

    const handleSearchInput = event => {
        setStock(event.target.value);
    }

    const submitInput = event => {
        if (event.key === 'Enter') {
            searchStock();
        }
    }

    function searchStock() {
        history.push(`/results/${stock}`);
    }

    return (
        <Wrapper>
            <Container>
                <TopBar/>
                <WrapperSearch>
                    <ContainerSearch>
                        <h1>Stock Market</h1>
                        <input type="text" placeholder="Search a Stock..." onChange={handleSearchInput} onKeyPress={submitInput} value={stock}/>
                        <button onClick={() => searchStock()}>Search</button>
                    </ContainerSearch>
                </WrapperSearch>
                
            </Container>
        </Wrapper>
  
    );
}

export default SearchPage;