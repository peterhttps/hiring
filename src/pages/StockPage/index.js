import React, { useEffect, useState } from "react";
import { useParams } from "react-router-dom";
import LoadingSpin from 'react-loading-spin';

import { Wrapper, Container, WrapperStock, ContainerStock, ContainerStockInfos, StockTitle, StockPrice } from "./styles";
import TopBar from "../../components/TopBar";

import api from '../../services/api';
import StockGraph from "../../components/StockGraph";
import TimeCompare from "../../components/TimeCompare";
import StockCompare from "../../components/StockCompare";
import StockAdd from "../../components/StockAdd";
import { LoadingSpinContainer } from "../../components/Global/styles";

function StockPage() {
    let { id } = useParams();
    const [loadingStockInfos, setLoadingStockInfos] = useState(true);
    const [stockInfos, setStockInfos] = useState([]);
    const [graphInfos, setGraphInfos] = useState([]);

    useEffect(() => {

        async function getStock() {
            const resultInterday = await api.get(`?function=GLOBAL_QUOTE&symbol=${id}&apikey=7MN1KILIT0TIGGLR`);
            const resultDaily = await api.get(`?function=TIME_SERIES_DAILY_ADJUSTED&symbol=${id}&outputsize=full&apikey=7MN1KILIT0TIGGLR`);
            const timeSeries = resultDaily.data["Time Series (Daily)"];
            
            setStockInfos(resultInterday.data["Global Quote"]); 
            setGraphInfos(timeSeries);
            setLoadingStockInfos(false);
            
        }

        getStock();

    }, [id]);

    function parseValueToFloat(string) {
        return parseFloat(string).toFixed(2)
    }

    return (
        <Wrapper>
            <Container>
                <TopBar />
                <WrapperStock>
                {!loadingStockInfos ?
                    <ContainerStock>
                        <StockTitle>{id}</StockTitle>
                         
                        <StockPrice>
                            <div>
                            {!loadingStockInfos && <h2> $ {parseValueToFloat(stockInfos["05. price"])}</h2>}
                            </div>
                        </StockPrice>
                        {!loadingStockInfos && <StockAdd actualStock={id} />}
                        <ContainerStockInfos>
                            <div>
                                <p>open</p>
                                {!loadingStockInfos && <h1>$ {parseValueToFloat(stockInfos["02. open"])}</h1>}
                            </div>
                            <div>
                                <p>high</p>
                                {!loadingStockInfos && <h1>$ {parseValueToFloat(stockInfos["03. high"])}</h1>}
                            </div>
                            <div>
                                <p>low</p>
                                {!loadingStockInfos && <h1>$ {parseValueToFloat(stockInfos["04. low"])}</h1>}
                            </div>
                            <div>
                                <p>volume</p>
                                {!loadingStockInfos && <h1>{stockInfos["06. volume"]}</h1>}
                            </div>
                            <div>
                                <p>latest trading day</p>
                                {!loadingStockInfos && <h1>{stockInfos["07. latest trading day"]}</h1>}
                            </div>
                        </ContainerStockInfos>
                        {!loadingStockInfos && <StockGraph data={graphInfos} />}
                        {!loadingStockInfos && <TimeCompare data={graphInfos} />}
                        {!loadingStockInfos && <StockCompare actualStockData={stockInfos} />}
                        

                       
                    </ContainerStock>
                     : <LoadingSpinContainer><LoadingSpin/></LoadingSpinContainer>}
                </WrapperStock>
            </Container>
        </Wrapper>
    );
}

export default StockPage;
