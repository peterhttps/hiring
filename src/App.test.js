import { render, screen } from "@testing-library/react";
import App from "./App";
import TestRenderer from "react-test-renderer"; // ES6

// Mock
import { mockData } from "./testsData/daily";

// Providers
import { StockProvider } from "./providers/stockProvider";

// Components
import StockAdd from "./components/StockAdd";
import TopBar from "./components/TopBar";
import StockGraph from "./components/StockGraph";
import TimeCompare from "./components/TimeCompare";

// Pages
import StockPage from "./pages/StockPage";

import { Router } from "react-router-dom";
import { createMemoryHistory } from "history";
import SearchPage from "./pages/SearchPage";

test("render SockGraph component", () => {
  const props = {
    data: mockData,
  };

  const Component = TestRenderer.create(<StockGraph {...props} />);
  const Tree = Component.toJSON();
  expect(Tree).toMatchSnapshot();
});

test("render TimeCompare component", () => {
  const props = {
    data: mockData,
  };

  const Component = TestRenderer.create(<TimeCompare {...props} />);
  const Tree = Component.toJSON();
  expect(Tree).toMatchSnapshot();
});

test("render StockAdd component", () => {
  const props = {
    actualStock: "IBM",
  };

  // Testing with context provider
  const Component = TestRenderer.create(
    <StockProvider>
      <StockAdd {...props} />
    </StockProvider>
  );
  const Tree = Component.toJSON();
  expect(Tree).toMatchSnapshot();
});

test("render TopBar component", () => {
  const Component = TestRenderer.create(<TopBar />);
  const Tree = Component.toJSON();
  expect(Tree).toMatchSnapshot();
});

test("render StockPage component", () => {
  const history = createMemoryHistory();
  const route = "/stock/IBM";
  history.push(route);
  const Component = TestRenderer.create(
    <Router history={history}>
      <StockPage />
    </Router>
  );
  const Tree = Component.toJSON();
  expect(Tree).toMatchSnapshot();
});


test("render SearchPage component", () => {
  const Component = TestRenderer.create(<SearchPage />);
  const Tree = Component.toJSON();
  expect(Tree).toMatchSnapshot();
});
