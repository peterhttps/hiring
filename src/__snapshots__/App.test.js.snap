// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render SearchPage component 1`] = `
<div
  className="sc-gsTCUz bCAtYS"
>
  <div
    className="sc-dlfnbm bFdpYG"
  >
    <div
      className="sc-jSgupP eYLJnf"
    >
      <button
        onClick={[Function]}
      >
        <svg
          color="#FFFFFF"
          fill="none"
          height={24}
          stroke="currentColor"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth="2"
          style={
            Object {
              "color": "#FFFFFF",
            }
          }
          viewBox="0 0 24 24"
          width={24}
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M21 12.79A9 9 0 1 1 11.21 3 7 7 0 0 0 21 12.79z"
          />
        </svg>
      </button>
      <div
        onClick={[Function]}
      >
        <h2>
          My Portfólio
        </h2>
      </div>
    </div>
    <div
      className="sc-hKgILt bOfHVR"
    >
      <div
        className="sc-eCssSg dfpGNO"
      >
        <h1>
          Stock Market
        </h1>
        <input
          onChange={[Function]}
          onKeyPress={[Function]}
          placeholder="Search a Stock..."
          type="text"
          value=""
        />
        <p>
          Example: ibm, microsoft, tesla
        </p>
        <button
          onClick={[Function]}
        >
          Search
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`render SockGraph component 1`] = `
<div
  className="sc-dmlrTW fBJuBq"
>
  <div
    className="sc-kfzAmx jOiRlD"
  >
    <p>
      From: 
    </p>
    <div
      className="react-datepicker-wrapper"
    >
      <div
        className="react-datepicker__input-container"
      >
        <button
          className="sc-bqyKva eRzpRF"
          onClick={[Function]}
        >
          2021-02-12
        </button>
      </div>
    </div>
    <p>
      To: 
    </p>
    <div
      className="react-datepicker-wrapper"
    >
      <div
        className="react-datepicker__input-container"
      >
        <button
          className="sc-bqyKva eRzpRF"
          onClick={[Function]}
        >
          2021-02-12
        </button>
      </div>
    </div>
    <button
      className="sc-crrsfI YSOOT"
      onClick={[Function]}
    >
      Filter
    </button>
  </div>
</div>
`;

exports[`render StockAdd component 1`] = `
<div
  className="sc-hiSbYr fYaTal"
>
  <button
    className="sc-crrsfI YSOOT"
    onClick={[Function]}
  >
    Add to Portfolio
  </button>
</div>
`;

exports[`render StockPage component 1`] = `
<div
  className="sc-hBEYos ftRKSd"
>
  <div
    className="sc-fodVxV emhgxZ"
  >
    <div
      className="sc-jSgupP eYLJnf"
    >
      <button
        onClick={[Function]}
      >
        <svg
          color="#FFFFFF"
          fill="none"
          height={24}
          stroke="currentColor"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth="2"
          style={
            Object {
              "color": "#FFFFFF",
            }
          }
          viewBox="0 0 24 24"
          width={24}
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="12"
            cy="12"
            r="5"
          />
          <line
            x1="12"
            x2="12"
            y1="1"
            y2="3"
          />
          <line
            x1="12"
            x2="12"
            y1="21"
            y2="23"
          />
          <line
            x1="4.22"
            x2="5.64"
            y1="4.22"
            y2="5.64"
          />
          <line
            x1="18.36"
            x2="19.78"
            y1="18.36"
            y2="19.78"
          />
          <line
            x1="1"
            x2="3"
            y1="12"
            y2="12"
          />
          <line
            x1="21"
            x2="23"
            y1="12"
            y2="12"
          />
          <line
            x1="4.22"
            x2="5.64"
            y1="19.78"
            y2="18.36"
          />
          <line
            x1="18.36"
            x2="19.78"
            y1="5.64"
            y2="4.22"
          />
        </svg>
      </button>
      <div
        onClick={[Function]}
      >
        <h2>
          My Portfólio
        </h2>
      </div>
    </div>
    <div
      className="sc-fFubgz fediFZ"
    >
      <div
        className="sc-dQppl kDMKwn"
      >
        <div
          className="loading"
          style={
            Object {
              "animationDirection": "normal",
              "animationDuration": "1.4s",
              "animationTimingFunction": "ease-in-out",
              "borderColor": "#cccccc",
              "borderLeftColor": "#1ecd96",
              "borderTopColor": "#1ecd96",
              "borderWidth": "6px",
              "height": "60px",
              "width": "60px",
            }
          }
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`render TimeCompare component 1`] = `
<div
  className="sc-fKFyDc fWXiWs"
>
  <h1>
    Time Compare
  </h1>
  <hr />
  <div
    className="sc-bBXqnf bVmiJm"
  >
    <div
      className="sc-iwyYcG dVMvYW"
    >
      <p>
        Purchase date: 
      </p>
      <div
        className="react-datepicker-wrapper"
      >
        <div
          className="react-datepicker__input-container"
        >
          <button
            className="sc-bqyKva eRzpRF"
            onClick={[Function]}
          >
            2021-02-12
          </button>
        </div>
      </div>
      <p>
        Sell date: 
      </p>
      <div
        className="react-datepicker-wrapper"
      >
        <div
          className="react-datepicker__input-container"
        >
          <button
            className="sc-bqyKva eRzpRF"
            onClick={[Function]}
          >
            2021-02-12
          </button>
        </div>
      </div>
      <button
        className="sc-crrsfI YSOOT"
        onClick={[Function]}
      >
        Calculate
      </button>
    </div>
  </div>
</div>
`;

exports[`render TopBar component 1`] = `
<div
  className="sc-jSgupP eYLJnf"
>
  <button
    onClick={[Function]}
  >
    <svg
      color="#FFFFFF"
      fill="none"
      height={24}
      stroke="currentColor"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeWidth="2"
      style={
        Object {
          "color": "#FFFFFF",
        }
      }
      viewBox="0 0 24 24"
      width={24}
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M21 12.79A9 9 0 1 1 11.21 3 7 7 0 0 0 21 12.79z"
      />
    </svg>
  </button>
  <div
    onClick={[Function]}
  >
    <h2>
      My Portfólio
    </h2>
  </div>
</div>
`;
